### ISSUE TITLE
Error Using the Facade

### ISSUE THREAD

**Issue by vinkla** [vinkla/laravel-pusher#2]
The error below has been reported by @alexandredes and @mottihoresh. This error occurs when using the Facade. As I've learned, using dependency injection it works.

``` php
call_user_func_array() expects parameter 1 to be a valid callback
```

I've tried this in my local setup, using Homestead, with both the Facade and dependency injection without any problems. The events gets registered in the console at http://pusher.com. Example code below.

``` php
<?php namespace App\Http\Controllers;

use Vinkla\Pusher\Facades\Pusher;

class PushController extends Controller {

    public function index()
    {
        dd(Pusher::trigger('my-channel', 'my-event', ['message' => 'A']));
    }

}
```

![unnamed](https://cloud.githubusercontent.com/assets/499192/6199405/2a4b724a-b442-11e4-9ba2-e683e6fd59ed.png)

If someone else experience this, please report it here. If you have a solution, please share it!

**Comment by a3gis** [comment]
Same problem here. No fix idea as of yet :/

**Comment by simderoemer** [comment]
Same issue.

**Comment by aydnbhdr** [comment]
It occurs on my project too.

**Comment by yhalabi** [comment]
Same here

**Comment by vinkla** [comment]
I think this is because how the Pusher library is built. The `Pusher` package currently autoloads everything in their `lib` directory. The `Pusher` class and `Pusher` facade might collide and overwrite each other.

I've have discussed a PSR-4 update to the library in https://github.com/pusher/pusher-http-php/pull/25. They are about to release a new package that is more up to date with PHP. Maybe we can hold out until then and release Laravel Pusher `2.0` when it is ready.

**Comment by Kryptonit3** [comment]
Any idea on when this will be fixed? Or what we can temporarily edit to make it work?

**Comment by mottihoresh** [comment]
I was having more issues it today. I am not too familiar with the way ServiceProviders work in Laravel, so i just created a very basic service injector

https://gist.github.com/mottihoresh/394e12ea802b41b5ef65

the only thing you need to do is to add pusher-php-server to your composer file.

**Comment by vinkla** [comment]
@mottihoresh @Kryptonit3 Have you tried using it as in our dependency injection example instead of using the facade?

**Comment by vinkla** [comment]
For reference how to use it with dependency injection you can check out this [RealTimeStatusHandler](https://github.com/StyleCI/StyleCI/blob/master/app/Handlers/Events/RealTimeStatusHandler.php) class by @StyleCI.

**Comment by TimothyLoyer** [comment]
I tried using the facade and then DI and get the same error (maybe since the PusherManager also ends up using the facade?).

```
[2015-03-17 20:21:43] local.ERROR: exception 'ErrorException' with message 'call_user_func_array() expects parameter 1 to be a valid callback, class 'Vinkla\Pusher\Facades\Pusher' does not have a method 'trigger'' in /home/vagrant/Code/l5-acme-api/vendor/graham-campbell/manager/src/AbstractManager.php:234
```

**Comment by vinkla** [comment]
@TimothyLoyer I think its because how the Pusher library is built, it currently return the Pusher class globally.

**Comment by TimothyLoyer** [comment]
Maybe having your facade named "Pusher" is causing an issue- might be why the L4 package used "pusherer".

**Comment by vinkla** [comment]
@TimothyLoyer Yes, I think so too. Maybe we can hold out on this until they release the new version https://github.com/pusher/pusher-http-php/tree/new-lib

**Comment by TimothyLoyer** [comment]
I just renamed the facade (and continued using DI, but I imagine that the facade also works now with its new name). The area of issue is in PusherFactory, where you have `use Pusher;` it is pulling in the your facade rather than the Pusher library.

**Comment by vinkla** [comment]
The facades in Laravel are globally registered so its quite hard to work around. Did it work renaming it?

**Comment by TimothyLoyer** [comment]
Yes, and you are right about the Pusher php library- all those classes are in the global namespace (ugly!) so that forces you into working around its namespaced classes. You'll want to double check all those classnames to be sure you don't have any overlap.

**Comment by vinkla** [comment]
@TimothyLoyer Yes, you can still use the `PusherManager` class without any problems.

**Comment by TimothyLoyer** [comment]
I tweeted Pusher about namespacing their composer package and got a pretty quick response from @leggetter. "We’ve got a new-lib branch which is an completely re-written lib. v3.0 will be the last non-namespaces major release." More info here: http://j.mp/1MKUTt3  - Looks like they are moving to a 4.0 branch.

**Comment by leggetter** [comment]
:+1:

**Comment by jadjoubran** [comment]
I can't seem to get it work with dependency injection as well.
Can you show an example how it works?

**Comment by vinkla** [comment]
@jadjoubran please see https://github.com/vinkla/pusher/issues/2#issuecomment-81042987. Yes, you could try to remove the facade.

**Comment by CoeusCC** [comment]
For the time being, I've just renamed my facade to `LaravelPusher` to great success. In my controller I call `use LaravelPusher;` then I can use 
```
public function pusher()
	{
		$messages = '{"name":"Joe","message":"Hello world!"}';
		LaravelPusher::trigger('test_channel', 'my_event', ['message' => $messages]);
	}
```
in my function. My alias then has to become `'LaravelPusher' => 'Vinkla\Pusher\Facades\Pusher',`.

As mentioned above, it's because the class and facade clash for the namespace.

**Comment by kunalthool** [comment]
Thanks that works like a charm.

**Comment by Kryptonit3** [comment]
here is a simple package - https://github.com/Kryptonit3/Pusher

**Comment by zlorkovic** [comment]
+1

**Comment by GrahamCampbell** [comment]
Just have `use Vinkla\Pusher\Facades\Pusher as PusherFacade;`. Then you're done...

**Comment by stephanedemotte** [comment]
any update ?

**Comment by GrahamCampbell** [comment]
> any update ?

There is no bug in this library.

**Comment by GrahamCampbell** [comment]
Go pester the actual pusher library to sort it out.

**Comment by sheyooo** [comment]
solution is here not to put the facade class in `app.php` and use the dependency injection with the service provider alone in the `app.php` file.....

**Comment by rulatir** [comment]
sheyooo - exact incantation please?

**Comment by sheyooo** [comment]
@rulatir  [https://github.com/vinkla/pusher#installation](url)
follow the instructions this way

## Installation
Require this package, with [Composer](https://getcomposer.org/), in the root directory of your project.

```bash
composer require vinkla/pusher
```

Add the service provider to `config/app.php` in the `providers` array.

```php
Vinkla\Pusher\PusherServiceProvider::class
```

~~If you want you can use the [facade](http://laravel.com/docs/facades). Add the reference in `config/app.php` to your aliases array.~~

```php
'Pusher' => Vinkla\Pusher\Facades\Pusher::class //IGNORE THIS
```

If you prefer to use dependency injection over facades like me, then you can inject the manager:

```php
use Vinkla\Pusher\PusherManager;

class Foo
{
	protected $pusher;

	public function __construct(PusherManager $pusher)
	{
		$this->pusher = $pusher;
	}

	public function bar()
	{
		$this->pusher->trigger('my-channel', 'my-event', ['message' => $message]);
	}
}

```

**Comment by rulatir** [comment]
Thanks sheyooo, I just ended up writing a super simple facade by hand, one that directly wraps the Pusher class. I removed vinkla/pusher altogether.

**Comment by vinkla** [comment]
@rulatir You can actually use the Pusher class with this package without any setup. That is why we bind the Pusher class to the connection. Please see [`PusherServiceProvider.php`](https://github.com/vinkla/pusher/blob/21a8311dca1b5040e2b2e8234da7b1f584e1f279/src/PusherServiceProvider.php#L103-L111).

**Comment by AymanAlMalki** [comment]
Using @GrahamCampbell suggestion solved the issue

**Comment by jrsalunga** [comment]
$pusher = app('pusher.connection');
 $pusher->trigger('test_channel', 'my_event', ['message'=>'hello world!']);

**Comment by mhdafiq** [comment]
I still don't know how this work. Tried all of the method.
Someone please guide me.

**Comment by vinkla** [comment]
@mhdafiq try https://github.com/vinkla/laravel-pusher/issues/2#issuecomment-179263896

**Comment by mhdafiq** [comment]
@vinkla i already tried that but still cannot trigger from server.

when i 
`dd($this->pusher->trigger('test_channel', 'my_event', ['message' => 'hello world']));`
the value return false.

**Comment by vinkla** [comment]
@jameshfisher @luismfonseca are there any updates on the new version of the `pusher/pusher-php-server` package? When will Pusher release the namespaced version?

**Comment by luismfonseca** [comment]
@GrahamCampbell @vinkla so I understand correctly, I can just add `namespace Pusher;` to the library, would this solve this problem?

Future users of the library would always need to do `use Pusher\Pusher;` right?

**Comment by vinkla** [comment]
Yes, but you need to release it as a new major version. It is a breaking change. Shouldn't be a problem though since integers are cheap.

**Comment by vinkla** [comment]
@luismfonseca let me know if there is anything I can help with!

**Comment by luismfonseca** [comment]
Btw, I've released the new php version!

**Comment by vinkla** [comment]
@luismfonseca great! I'll release a new version of this package with the Laravel 5.5 support release.

**Comment by normanaranez** [comment]
ca we use facade now?

**Comment by vinkla** [comment]
I'll release a new version soon.

**Comment by vinkla** [comment]
Fixed in 25a88e4be077218e890b995afe45e8854ef6865f.
